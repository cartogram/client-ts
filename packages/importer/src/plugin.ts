import { XataPlugin, XataPluginOptions } from '@xata.io/client';
import { Importer } from './importer';
import { ImportCsvOptions, ImportFileOptions, ImportJsonOptions, ImportNdJsonOptions } from './types';

export class XataImportPlugin extends XataPlugin {
  build(_options: XataPluginOptions) {
    const importer = new Importer();

    return {
      file: (options: Omit<ImportFileOptions, 'strategy'>) => importer.read({ strategy: 'file', ...options }),
      json: (options: Omit<ImportJsonOptions, 'strategy'>) => importer.read({ strategy: 'json', ...options }),
      ndjson: (options: Omit<ImportNdJsonOptions, 'strategy'>) => importer.read({ strategy: 'ndjson', ...options }),
      csv: (options: Omit<ImportCsvOptions, 'strategy'>) => importer.read({ strategy: 'csv', ...options })
    };
  }
}
