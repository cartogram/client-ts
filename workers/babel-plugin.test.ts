import babel from '@babel/core';
import { afterEach, describe, expect, test, vi } from 'vitest';
import plugin from './babel-plugin.js';
import fs from 'fs';

vi.mock('fs');

const writeFileSyncMock = fs.writeFileSync as unknown as ReturnType<typeof vi.fn>;

const options = { clientPath: 'xata.ts', output: '/output', root: '/client' };

afterEach(() => {
  vi.clearAllMocks();
});

describe('babel plugin', () => {
  test('no xata workers', async () => {
    const code = `
var foo = 1;
if (foo) console.log(foo);
`;

    babel.transformSync(code, { filename: 'source.js', plugins: [[plugin, options]] });

    expect(writeFileSyncMock.mock.calls[0]).toMatchSnapshot();
  });

  test('a simple xata worker', async () => {
    const code = `

xataWorker(async function helloWorld() {
  return 'Hello world';
})

`;

    babel.transformSync(code, { filename: 'source.js', plugins: [[plugin, options]] });

    expect(writeFileSyncMock.mock.calls[0]).toMatchSnapshot();
  });

  test('a xata worker with arguments', async () => {
    const code = `

xataWorker(async function paginatedResult(page) {
  return page;
})

`;

    babel.transformSync(code, { filename: 'source.js', plugins: [[plugin, options]] });

    expect(writeFileSyncMock.mock.calls[0]).toMatchSnapshot();
  });

  test('a xata worker with an external module dependency', async () => {
    const code = `
import camelCase from 'camelcase';

xataWorker(async function camelCasedHelloWorld() {
  return camelCase('Hello world');
})

`;

    babel.transformSync(code, { filename: 'source.js', plugins: [[plugin, options]] });

    expect(writeFileSyncMock.mock.calls[0]).toMatchSnapshot();
  });

  test('a xata worker with a local dependency', async () => {
    const code = `
import dep from './dep.js';

xataWorker(async function localDependency() {
  return dep.doSomething();
})

`;

    babel.transformSync(code, { filename: 'source.js', plugins: [[plugin, options]] });

    expect(writeFileSyncMock.mock.calls[0]).toMatchSnapshot();
  });
});
