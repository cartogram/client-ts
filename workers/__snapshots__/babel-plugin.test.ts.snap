// Vitest Snapshot v1

exports[`babel plugin > a simple xata worker 1`] = `
[
  "/output/index.js",
  "


import { XataClient } from '/client/xata.ts'

const corsHeaders = {
'Access-Control-Allow-Origin': '*',
'Access-Control-Allow-Methods': 'GET,HEAD,POST,OPTIONS',
'Access-Control-Max-Age': '86400',
};

const workers = {
'helloWorld': async function helloWorld() {
  return 'Hello world';
}
}

export default {
async fetch(request) {
  if (request.method === 'OPTIONS') {
    return new Response(null, {
      headers: corsHeaders,
    })
  } else {
    const url = new URL(request.url)
    const worker = url.searchParams.get('worker')
    const args = JSON.parse(url.searchParams.get('args'))
    const result = await workers[worker].apply({
      xata: new XataClient({apiKey: XATA_API_KEY }),
    }, args)
    return new Response(JSON.stringify(result), {
      headers: corsHeaders
    });
  }
},
};

",
]
`;

exports[`babel plugin > a xata worker with a local dependency 1`] = `
[
  "/output/index.js",
  "
import $dep_ec16d01 from '/Users/gimenete/projects/github.com/xataio/client-ts/dep.js'

import { XataClient } from '/client/xata.ts'

const corsHeaders = {
'Access-Control-Allow-Origin': '*',
'Access-Control-Allow-Methods': 'GET,HEAD,POST,OPTIONS',
'Access-Control-Max-Age': '86400',
};

const workers = {
'localDependency': async function localDependency() {
  return $dep_ec16d01.doSomething();
}
}

export default {
async fetch(request) {
  if (request.method === 'OPTIONS') {
    return new Response(null, {
      headers: corsHeaders,
    })
  } else {
    const url = new URL(request.url)
    const worker = url.searchParams.get('worker')
    const args = JSON.parse(url.searchParams.get('args'))
    const result = await workers[worker].apply({
      xata: new XataClient({apiKey: XATA_API_KEY }),
    }, args)
    return new Response(JSON.stringify(result), {
      headers: corsHeaders
    });
  }
},
};

",
]
`;

exports[`babel plugin > a xata worker with an external module dependency 1`] = `
[
  "/output/index.js",
  "
import $camelCase_8cea4cd from '/client/node_modules/camelcase'

import { XataClient } from '/client/xata.ts'

const corsHeaders = {
'Access-Control-Allow-Origin': '*',
'Access-Control-Allow-Methods': 'GET,HEAD,POST,OPTIONS',
'Access-Control-Max-Age': '86400',
};

const workers = {
'camelCasedHelloWorld': async function camelCasedHelloWorld() {
  return $camelCase_8cea4cd('Hello world');
}
}

export default {
async fetch(request) {
  if (request.method === 'OPTIONS') {
    return new Response(null, {
      headers: corsHeaders,
    })
  } else {
    const url = new URL(request.url)
    const worker = url.searchParams.get('worker')
    const args = JSON.parse(url.searchParams.get('args'))
    const result = await workers[worker].apply({
      xata: new XataClient({apiKey: XATA_API_KEY }),
    }, args)
    return new Response(JSON.stringify(result), {
      headers: corsHeaders
    });
  }
},
};

",
]
`;

exports[`babel plugin > a xata worker with arguments 1`] = `
[
  "/output/index.js",
  "


import { XataClient } from '/client/xata.ts'

const corsHeaders = {
'Access-Control-Allow-Origin': '*',
'Access-Control-Allow-Methods': 'GET,HEAD,POST,OPTIONS',
'Access-Control-Max-Age': '86400',
};

const workers = {
'paginatedResult': async function paginatedResult(page) {
  return page;
}
}

export default {
async fetch(request) {
  if (request.method === 'OPTIONS') {
    return new Response(null, {
      headers: corsHeaders,
    })
  } else {
    const url = new URL(request.url)
    const worker = url.searchParams.get('worker')
    const args = JSON.parse(url.searchParams.get('args'))
    const result = await workers[worker].apply({
      xata: new XataClient({apiKey: XATA_API_KEY }),
    }, args)
    return new Response(JSON.stringify(result), {
      headers: corsHeaders
    });
  }
},
};

",
]
`;

exports[`babel plugin > no xata workers 1`] = `
[
  "/output/index.js",
  "


import { XataClient } from '/client/xata.ts'

const corsHeaders = {
'Access-Control-Allow-Origin': '*',
'Access-Control-Allow-Methods': 'GET,HEAD,POST,OPTIONS',
'Access-Control-Max-Age': '86400',
};

const workers = {

}

export default {
async fetch(request) {
  if (request.method === 'OPTIONS') {
    return new Response(null, {
      headers: corsHeaders,
    })
  } else {
    const url = new URL(request.url)
    const worker = url.searchParams.get('worker')
    const args = JSON.parse(url.searchParams.get('args'))
    const result = await workers[worker].apply({
      xata: new XataClient({apiKey: XATA_API_KEY }),
    }, args)
    return new Response(JSON.stringify(result), {
      headers: corsHeaders
    });
  }
},
};

",
]
`;

exports[`babel plugin > supports TypeScrript 1`] = `
[
  "/output/index.js",
  "


import { XataClient } from '/client/xata.ts'

const corsHeaders = {
'Access-Control-Allow-Origin': '*',
'Access-Control-Allow-Methods': 'GET,HEAD,POST,OPTIONS',
'Access-Control-Max-Age': '86400',
};

const workers = {

}

export default {
async fetch(request) {
  if (request.method === 'OPTIONS') {
    return new Response(null, {
      headers: corsHeaders,
    })
  } else {
    const url = new URL(request.url)
    const worker = url.searchParams.get('worker')
    const args = JSON.parse(url.searchParams.get('args'))
    const result = await workers[worker].apply({
      xata: new XataClient({apiKey: XATA_API_KEY }),
    }, args)
    return new Response(JSON.stringify(result), {
      headers: corsHeaders
    });
  }
},
};

",
]
`;
